### PR TITLE
Sort preserving order for composite keys! Woohoo!

### PR DESCRIPTION
**Pull request by erights**
`encodeKey` now works for composite keys (copyArray, copyRecord, tagged) and therefore for all key types (copySet, copyBag, copyMap). It produces strings that are sort order preserving. See `orderInvariants` for a number of invariants between these orders that are passing, at least on the current test data.

### CODE REVIEW

File: packages/store/src/patterns/encodeKey.js [hunk 1]
```diff
@@ -140,6 +144,94 @@ const dbEntryKeyToBigint = k => {
   }
 };
 
+// `'\u0000'` is the terminator after elements.
+// `'\u0001'` is the backslash-like escape character, for
+// escaping both of these characters.
```
Comment (FUDCo) [inline packages/store/src/patterns/encodeKey.js:149]: Do we have a suitable reservoir of Unicode pedantry in the house to be sure this encoding doesn't alias to something else weird if these codes precede or follow any other value?

**Review by FUDCo — APPROVED** [review]
Very clever!  I do have one concern raised in a comment.  I'm betting it's not a problem but I'd like somebody more versed in Unicode funkiness to weigh in (it's entirely possible that your own knowledge of Unicode is sufficient, but mine definitely is not).

File: packages/store/src/patterns/encodeKey.js (hunk 1 above)
Comment (erights) [reply packages/store/src/patterns/encodeKey.js:149]: Yeah, I'm confident enough.

File: packages/store/src/patterns/rankOrder.js [hunk 2]
```diff
@@ -66,6 +71,18 @@ const memoOfSorted = new WeakMap();
  */
 const comparatorMirrorImages = new WeakMap();
 
+export const recordParts = record => {
+  assertRecord(record);
+  // TODO Measure which is faster: a reverse sort by sorting and
+  // reversing, or by sorting with an inverse comparison function.
+  // If it makes a significant difference, use the faster one.
+  const names = ownKeys(record).sort().reverse();
+  // @ts-expect-error It thinks name might be a symbol, which it doesn't like.
```
Comment (mhofman) [inline packages/store/src/patterns/rankOrder.js:80]: We really need to have generic types for `ownKeys` and co.

File: packages/store/src/patterns/encodeKey.js [hunk 3]
```diff
@@ -140,6 +144,94 @@ const dbEntryKeyToBigint = k => {
...
+  }
+  return chars.join('');
+};
+
+const decodeArray = (encodedKey, decodeKey) => {
+  assert(encodedKey.startsWith('['), X`Encoded array expected: ${encodedKey}`);
+  const elements = [];
+  const elemChars = [];
+  for (let i = 1; i < encodedKey.length; i += 1) {
+    const c = encodedKey[i];
```
Comment (mhofman) [inline packages/store/src/patterns/encodeKey.js:171]: This will break codepoints into multiple code units. Why not `for (const c of encodedKey)` like encode does it?

File: packages/store/test/test-rankOrder.js [hunk 4]
```diff
@@ -120,23 +120,6 @@ const sortedSample = harden([
   // Lexicographic arrays. Shorter beats longer.
   // Lexicographic records by reverse sorted property name, then by values
   // in that order.
-
```
Comment (mhofman) [inline packages/store/test/test-rankOrder.js:123]: We need some multi code unit chars in these tests to make sure we're unicode safe.

**Review by mhofman — COMMENTED** [review]
I believe there is a unicode handling issue in decodeArray with characters spanning 2 code units.

File: packages/store/src/patterns/rankOrder.js (hunk 2 above)
Comment (erights) [reply packages/store/src/patterns/rankOrder.js:80]: I don't understand. What is generic or generic-able about the type of `ownKeys`?

File: packages/store/src/patterns/encodeKey.js (hunk 3 above)
Comment (erights) [reply packages/store/src/patterns/encodeKey.js:171]: > This will break codepoints into multiple code units.

How does this create a problem? Can you construct an example where it does the wrong thing?

> Why not for (const c of encodedKey) like encode does it?

because I need to skip ahead one inside the loop. And it would still iterate code units anyway.

File: packages/store/test/test-rankOrder.js (hunk 4 above)
Comment (erights) [reply packages/store/test/test-rankOrder.js:123]: PR please with such tests? I'll happily review.

File: packages/store/src/patterns/rankOrder.js (hunk 2 above)
Comment (mhofman) [reply packages/store/src/patterns/rankOrder.js:80]: `assertRecord` narrows the type of `record` to `CopyRecord<any>` which means `record: Record<string, any>`, and no keys would be symbols (or number).

The problem is that the type of `ownKeys` is `(target: object) => (string | symbol)[]`. If it was generic, aka `ownKeys<K extends string | symbol>(target: Record<K, any>) => K[]`, we wouldn't have a need for this `@ts-expect-error`, as the type of keys would flow through.

File: packages/store/src/patterns/encodeKey.js (hunk 3 above)
Comment (mhofman) [reply packages/store/src/patterns/encodeKey.js:171]: `for (const c of encodedKey)` iterates code points not code units. It would be possible to substitute the logic of skip ahead by using a flag indicating if the next char needs to be processed as an escaped char.

That said given the simple logic of escaping here and the fact low surrogates cannot be in the range of the escaped chars, there should be no risk in this decode. It's just usually best practice to not handle code units and only deal with code points.

File: packages/store/test/test-rankOrder.js (hunk 4 above)
Comment (mhofman) [reply packages/store/test/test-rankOrder.js:123]: I was thinking something along the lines of [Mathias' Pile of Poo Test](https://mathiasbynens.be/notes/javascript-unicode#poo-test): `Iñtërnâtiônàlizætiøn☃💩`
